### PR TITLE
WIP: Provide access to result of previous calculation

### PR DESCRIPTION
**Pull request by leftshift**
As discussed in #104, this adds access to the result of the previous operation via `ans`, `ANS` and `_`.

Was a bit harder than anticipated, as the results are converted to `NumberParts` before passed as a `QueryReply`, which doesn't contain the raw number result. As a workaround, I added the `raw_number` field to `NumberParts`, but it might make sense to instead convert to `NumberParts` at the very end instead?

To satisfy the `derive(Serialize, Deserialize)` on `NumberParts`, I had to make a few more things have these traits. I chose to enable the serde feature of the num crate and removed the custom serialize impl for BigRat. I'm not sure whether the NumberParts are actually serialized somewhere? And whether this change might break things there? 

Also, this seems to have broken something with the unit canonicalisation:
```
> 10 ohm
10 ohm (resistance)
> _ * 100mA
1 ampereohm
> 10 ohm * 100mA
1 volt (electrical_potential)
```
Haven't quite tracked down where that goes wrong.

What do you think? I'd appreciate feedback on what I already did and how best to continue. Thanks!

### CODE REVIEW

**Comment by leftshift** [comment]
Also, right now, when there is no previous result, `ans` behaves like a unit that doesn't exist (because lookup returns None). I think that could be pretty confusing, but changing this would require either changing the parser or how lookup works/what it can return I think.

File: core/src/context.rs [hunk 1]
```diff
@@ -82,6 +84,9 @@ impl Context {
     /// prefixes, plurals, bare dimensions like length, and quantities.
     pub fn lookup(&self, name: &str) -> Option<Number> {
         fn inner(ctx: &Context, name: &str) -> Option<Number> {
+            if name.eq("ans") {
```
Comment (tiffany352) [inline core/src/context.rs:87]:
```suggestion
            if name == "ans" || name == "ANS" || name == "_" {
```

File: core/src/lib.rs [hunk 2]
```diff
@@ -78,7 +78,14 @@ pub fn eval(ctx: &mut Context, line: &str) -> Result<QueryReply, QueryError> {
     ctx.update_time();
     let mut iter = text_query::TokenIterator::new(line.trim()).peekable();
     let expr = text_query::parse_query(&mut iter);
-    ctx.eval_outer(&expr)
+    let res = ctx.eval_outer(&expr)?;
+    if let QueryReply::Number(ref number_parts) = res {
+        println!("got number {}", number_parts);
```
Comment (tiffany352) [inline core/src/lib.rs:83]:
```suggestion
```

**Review by tiffany352 — APPROVED** [review]
Thanks for working on this!

Longer term I'd like to refactor the *Reply objects to avoid using NumberParts, but for the time being this should be fine.

I had some minor tweaks I'd like to make before merging but it otherwise looks good to me.

**Comment by leftshift** [comment]
Yay, thanks for the feedback!

Oops, did pretty much exactly those changes already but I forgot committing them it seems, thanks!

I think there still are some todos before merging:
- [ ] figure out what goes wrong with the unit canonicalisation (as mentioned in the pull request description)
- [ ] add documentation to the manual

I won't get around to working on that until in a few days probably.

**Comment by leftshift** [comment]
I think I understand the unit problem: The stored result isn't in base units, and by returning the display units from `lookup` like this, they incorrectly get treated as base units instead of being resolved back to base units. I'm not sure how this can be fixed, the cleanest way would probably be to try to also keep the Number in base units around when prettifying the result and store that?

**Comment by leftshift** [comment]
Okay, phew, got it, shouldn't have used the prettified value in `to_parts`. The units seem to work now, but I'm not 100% confident there isn't any other edge case where the units might go wrong.
Turns out you can't really do pull requests for github wikis, but I guess I'll write something up and attach a patch to this discussion.

**Comment by leftshift** [comment]
Patch for the wiki:
[0001-add-section-about-previous-result.txt](https://github.com/tiffany352/rink-rs/files/7022729/0001-add-section-about-previous-result.txt)

Feel free to reword or move to a different section, I wasn't really sure what's the best place for it in the manual.

**Comment by tiffany352** [comment]
The patch to the wiki look good to me. Thanks!

**Pull request merged**